golang/cosmos/x/swingset/types/msgs.go

```diff
@@ -183,3 +184,32 @@ func (msg MsgProvision) GetSignBytes() []byte {
 func (msg MsgProvision) GetSigners() []sdk.AccAddress {
 	return []sdk.AccAddress{msg.Submitter}
 }
+
+func NewMsgInstallBundle(bundleJson string, submitter sdk.AccAddress) *MsgInstallBundle {
+	return &MsgInstallBundle{
+		Bundle:    bundleJson,
+		Submitter: submitter,
+	}
+}
+
+// Route should return the name of the module
+func (msg MsgInstallBundle) Route() string { return RouterKey }
+
+// Type should return the action
+func (msg MsgInstallBundle) Type() string { return "installBundle" }
+
+// ValidateBasic runs stateless checks on the message
+func (msg MsgInstallBundle) ValidateBasic() error {
+	if msg.Submitter.Empty() {
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidAddress, "Submitter address cannot be empty")
+	}
+	if len(msg.Bundle) == 0 {
```

Comment:
I'm guessing it's not sufficiently convenient+useful to check e.g. JSON-shape, expected keys, base64-ness, or deeper?

I suppose the only benefit of additional checks would be that I might pay for an installation of something bogus, and get it into a block, only to get rejected after that block was executed by validators.

Eventually I think we want to do more extensive checks here to prevent abuse of blocks as a storage mechanism, catch the misshapen junk before it can get added to a block or flooded to all nodes.

Reply:
I believe keeping the bundle format opaque at this layer keeps with the spirit of Michael’s intention for the type to evolve freely.